docker/docker_client.go

```diff
@@ -253,11 +260,37 @@ func newDockerClient(ctx *types.SystemContext, ref dockerReference, write bool,
 		signatureBase: sigBase,
 		scope: authScope{
 			actions:    actions,
-			remoteName: reference.Path(ref.ref),
+			remoteName: remoteName,
 		},
 	}, nil
 }
 
+func getAuth(ctx *types.SystemContext, registry string) (string, string, error) {
+	if ctx != nil && ctx.DockerAuthConfig != nil {
+		return ctx.DockerAuthConfig.Username, ctx.DockerAuthConfig.Password, nil
+	}
+	return config.GetAuthentication(registry)
+}
+
+// CheckAuth validates the credentials by attempting to log into the registry
+func CheckAuth(ctx context.Context, sCtx *types.SystemContext, username, password, registry string) (bool, error) {
+	registry = config.DockerHostNameToRegistry(registry)
+	newLoginClient, err := newDockerClientWithDetails(sCtx, dockerReference{}, registry, username, password, "", false)
+	if err != nil {
+		return false, errors.Wrapf(err, "error creating new docker client")
+	}
+
+	resp, err := newLoginClient.makeRequest(ctx, "GET", "/v2/", nil, nil)
+	if err != nil {
+		return false, errors.Errorf("error occured with status code %q", resp.StatusCode)
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		return true, nil
+	}
+	return false, nil
```

Comment:
Please at least document when the function returns `true`/`false`.  Or, preferably, name it `AreCredentialsValid` or something like that, to make the meaning of the boolean obvious at the call site as well, having to read neither the implementation nor the documentation of `CheckAuth`.

Reply:
fixed

Reply:
Actually, when does the `false, nil` case occur? If I am reading the code correctly, when using `Authentication: Bearer`, failure is reported as `false, ErrUnauthorized`, but for `Authentication: Basic`, failure is reported as `false, nil`. It would be nice to be consistent.

Reply:
fixed

Reply:
AFAICS this is still inconsistent (and undocumented). The code now says `return false, err` instead of `return false, nil`, but `err` is guaranteed to be `nil` on that path.

Reply:
How would you suggest I restructure it? Should I document it better or change the code around?

Reply:
This function should probably end with a similar same `switch res.StatusCode` which is in `getBearerToken` (distinguishing between 200/401/unexpected).

Then there are two options:
- Define the "incorrect password" return value to be `false, nil`; in that case the existing `if err == ErrUnauthorized` conditional needs to be modified – and `ErrUnauthorized` can be made private.
- Define the "incorrect password" return value to be `false, ErrUnauthorized`; in that case the boolean return value is entirely equal to `err == nil`, and should be dropped entirely.

Either would work fine, the second one seems a tiny bit simpler.

Reply:
So like this? Since makeRequest will only return `err == nil` when the response status code is 200.
```
resp, err := newLoginClient.makeRequest(ctx, "GET", "/v2/", nil, nil)
	if err == ErrUnauthorized {
		return false, err
	}
	if err != nil {
		return false, errors.Errorf("error occured with status code %q", resp.StatusCode)
	}
	return true, nil
```

Reply:
> Since makeRequest will only return `err == nil` when the response status code is 200.

Really? https://golang.org/pkg/net/http/#Client.Do:
> A non-2xx status code doesn't cause an error.

`getBearerToken` is also aborting on `err` before checking `res.StatusCode`.

Am I overlooking something?

Reply:
Do note that there are two authentication mechanisms: `Authorization: Bearer`, and `Authorization: Basic`, depending on `c.challenges`.  `ErrUnauthorized` can only happen on the `Bearer` path, via `getBearerToken`; on the `Basic` path, nothing will create an `ErrUnauthorized` with the current code.

Reply:
From what I see `getBearerToken` returns an error when the code is not either a 200 or 401. If the code is 401, the `ErrUnauthorized` is return, otherwise a wrapped error is returned.
So if `getBearerToken` fails, `setUpRequestAuthFails`, which then causes `makeRequest` to fail with an error. And I am checking for the ErrUnauthorized case and any other error.
That is why I thought if makeRequest doesn't return an error then the authentication succeeded.
I might be reading the code wrong also.

Reply:
> So if getBearerToken fails, setUpRequestAuthFails, which then causes makeRequest to fail with an error.

That’s all true, but if the code uses `case "basic"`, `getBearerToken` is never called, and the authentication failure would (presumably) show up as a 401 only in the final response returned by `makeRequest`.

Reply:
oh okay, would this be better?
```
resp, err := newLoginClient.makeRequest(ctx, "GET", "/v2/", nil, nil)
	if err != nil {
		return false, err
	}
	switch resp.StatusCode {
	case http.StatusOK:
		return true, nil
	case http.StatusUnauthorized:
		return false, ErrUnauthorized
	default:
		return false, errors.Errorf("error occured with status code %q", resp.StatusCode)
	}
```

Reply:
Yes (if it works — I didn’t test it).

The boolean return value is now equivalent to `err == nil`, and can be entirely removed.